File: pricing-simulation/src/agent.ts
import "dotenv/config";
import { getEnvOrThrow } from "./utils.js";
import { Payments, type EnvironmentName } from "@nevermined-io/payments";
import OpenAI from "openai";

const OPENAI_API_KEY = getEnvOrThrow("OPENAI_API_KEY");
const NVM_API_KEY = getEnvOrThrow("NVM_API_KEY");
const NVM_ENVIRONMENT = getEnvOrThrow("NVM_ENVIRONMENT");

const payments = Payments.getInstance({
  nvmApiKey: NVM_API_KEY,
  environment: NVM_ENVIRONMENT as EnvironmentName,
});

const agent = async () => {
  console.log("Starting simulation request...");
  const agentRequest = await payments.requests.startSimulationRequest();

  console.log("Making OpenAI LLM call...");
  const openai = new OpenAI(
    payments.observability.withOpenAI(OPENAI_API_KEY, agentRequest, {})
  );

  const completion = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [{ role: "user", content: "Hello, world!" }],
  });

  // redeem credits
  console.log("Redeeming credits...");
  await payments.requests.finishSimulationRequest(agentRequest.agentRequestId);

  return completion.choices[0]?.message?.content;
};

const main = async () => {
  console.log("Calling agent...");
  const response = await agent();
  console.log(response);
};

main().then(() => {
  process.exit(0);
});

File: pricing-simulation-py/src/agent.py
import os
import sys
from dataclasses import asdict
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables from .env file BEFORE importing payments_py
# This is important because payments_py reads environment variables at import time
load_dotenv()

import openai
from payments_py import Payments, PaymentOptions

# Add the src directory to the path so we can import utils
sys.path.insert(0, str(Path(__file__).parent))
from utils import get_env_or_throw


def main():
    """Main function to run the pricing simulation agent."""
    
    # Load environment variables
    OPENAI_API_KEY = get_env_or_throw("OPENAI_API_KEY")
    NVM_API_KEY = get_env_or_throw("NVM_API_KEY")
    NVM_ENVIRONMENT = get_env_or_throw("NVM_ENVIRONMENT")

    # Initialize the Payments instance
    payments = Payments.get_instance(
        PaymentOptions(
            nvm_api_key=NVM_API_KEY,
            environment=NVM_ENVIRONMENT,
        )
    )

    print("Calling agent...")
    
    # Start simulation request
    print("Starting simulation request...")
    agent_request = payments.requests.start_simulation_request()

    # Make OpenAI LLM call
    print("Making OpenAI LLM call...")
    openai_config = payments.observability.with_openai(
        OPENAI_API_KEY,
        agent_request,
        {}
    )
    
    openai_client = openai.OpenAI(**asdict(openai_config))
    
    completion = openai_client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": "Hello, world!"}],
    )

    # Redeem credits
    print("Redeeming credits...")
    payments.requests.finish_simulation_request(agent_request.agent_request_id)

    # Return the response
    response = completion.choices[0].message.content if completion.choices else None
    print(response)


if __name__ == "__main__":
    main()

